### PR TITLE
Master post mortem crash manager adt

### PR DESCRIPTION
**Pull request by Elkasitu**
- [x] add description option, to be injected into the CrashManager's body
- [x] rename `name` to `label`
- [x] rename `act_xmlid` field to `action`
- [x] ~~log deactivated automated actions~~
- [x] tests
- [x] simplify spec

### CODE REVIEW

File: addons/base_automation/models/base_automation.py [hunk 1]
```diff
@@ -26,6 +26,20 @@
 }
 
 
+class AutomationFailureWizard(models.TransientModel):
+    _name = 'base.automation.failure.wizard'
+    _description = 'Automated action failure wizard'
+
+    def deactivate(self):
```
Comment (Elkasitu) [inline addons/base_automation/models/base_automation.py]: @api.model

File: addons/web/static/src/js/core/dialog.js [hunk 2]
```diff
@@ -65,9 +67,11 @@ var Dialog = Widget.extend({
             $content: false,
             buttons: [{text: _t("Ok"), close: true}],
             technical: true,
+            description: '',
         });
 
         this.$content = options.$content;
+        this.$content.append('<br/><br/>' + options.description);
```
Comment (Elkasitu) [inline addons/web/static/src/js/core/dialog.js]: sanitized?

File: addons/web/static/src/js/core/dialog.js [hunk 3]
```diff
@@ -65,9 +66,13 @@ var Dialog = Widget.extend({
...
             buttons: [{text: _t("Ok"), close: true}],
             technical: true,
+            description: '',
         });
 
-        this.$content = options.$content;
+        this.$content = options.$content
+        if (options.description) {
+            this.$content.append('<br/><br/>' + options.description);
+        }
```
Comment (Elkasitu) [inline addons/web/static/src/js/core/dialog.js]: probably not the best way to do this?

File: addons/web/static/src/js/services/crash_manager.js [hunk 4]
```diff
@@ -129,24 +129,47 @@ var CrashManager = core.Class.extend({
...
+        }
+        return btns;
+    },
     show_warning: function(error) {
         if (!this.active) {
             return;
         }
-        return new Dialog(this, {
+        var meta = error.data.metadata;
+        var dialog =  new Dialog(this, {
```
Comment (Elkasitu) [inline addons/web/static/src/js/services/crash_manager.js]: whoops, undo

**Review by VincentSchippefilt — COMMENTED** [review]
I reviewed the javascript parts, it's OK on the principle and requires minor changes

File: addons/web/static/src/js/services/crash_manager.js [hunk 5]
```diff
@@ -129,24 +130,47 @@ var CrashManager = core.Class.extend({
             this.show_error(error);
         }
     },
+    getMetaBtns: function (metadata) {
```
[the remark this answers is not in the document]
Comment (Elkasitu) [reply addons/web/static/src/js/services/crash_manager.js]: I agree with you on the metadata not meaning anything, it was originally called a 'post-mortem action' but @antonylesuisse insisted on it being called something like 'metadata' because I think he wants it to be generic for future uses...

Mind you this is not for server actions only, so I would call it simply 'action'

[the remark this answers is not in the document]
Comment (VincentSchippefilt) [reply addons/web/static/src/js/services/crash_manager.js]: `actions` is nice :)
if in the future it needs to be extended to more stuff, we will call it something less specific (like `stuff`)

**Comment by Elkasitu** [comment]
@VincentSchippefilt Let me know if it's OK now and I'll start writing some JS tests

File: odoo/exceptions.py [hunk 6]
```diff
@@ -13,24 +13,31 @@
...
             _logger.warn('except_orm is deprecated. Please use specific exceptions like UserError or AccessError. Caller: %s:%s', *caller)
         self.name = name
         self.value = value
         self.args = (name, value)
 
+        # only send the action to the web client if the user has the right permissions
+        if action and action.get('visible', True):
+            self.action = action
+        else:
+            self.action = False
```
Comment (rco-odoo) [inline odoo/exceptions.py]: The permission is defined where the action is defined.
This code processes something that says: "*I made some effort to compute something that is useless.*"

File: odoo/exceptions.py [hunk 7]
```diff
@@ -13,24 +13,31 @@
...
 from .tools.func import frame_codeinfo
+from .tools.translate import _
 
 _logger = logging.getLogger(__name__)
 
 
 # kept for backward compatibility
 class except_orm(Exception):
-    def __init__(self, name, value=None):
+    def __init__(self, name, value=None, action=None):
```
Comment (rco-odoo) [inline odoo/exceptions.py]: What about having several possible actions?

File: odoo/exceptions.py [hunk 8]
```diff
@@ -107,3 +114,53 @@ def __init__(self, msg, tb):
 
 class QWebException(Exception):
     pass
+
+
+class ViewError(except_orm, ValueError):
```
Comment (rco-odoo) [inline odoo/exceptions.py]: I would remove the inheritance with `ValueError`.
I don't think it brings anything useful.

File: odoo/exceptions.py [hunk 9]
```diff
@@ -107,3 +114,53 @@ def __init__(self, msg, tb):
...
+        ) % {
+            'view_name': view.name or not_avail,
+            'viewid': view.id or not_avail,
+            'xmlid': view.xml_id or not_avail,
+            'model': view.model or not_avail,
+            'parent': view.inherit_id.id or not_avail,
+            'msg': msg,
+        }
+        vals = [self, message, '']
+        _logger.info(message)
```
Comment (rco-odoo) [inline odoo/exceptions.py]: An `Exception` is essentially a data structure.
Logging in its initializer is inappropriate IMHO: you can instantiate an exception without raising it.

File: odoo/exceptions.py [hunk 10]
```diff
@@ -107,3 +114,53 @@ def __init__(self, msg, tb):
...
+            "model: %(model)s, parent_id: %(parent)s]"
+        ) % {
+            'view_name': view.name or not_avail,
+            'viewid': view.id or not_avail,
+            'xmlid': view.xml_id or not_avail,
+            'model': view.model or not_avail,
+            'parent': view.inherit_id.id or not_avail,
+            'msg': msg,
+        }
+        vals = [self, message, '']
```
Comment (rco-odoo) [inline odoo/exceptions.py]: Putting all that data in the error message is useless if the error provides a direct link to the view.

File: odoo/http.py [hunk 11]
```diff
@@ -703,6 +704,8 @@ def serialize_exception(e):
         tmp["exception_type"] = "access_denied"
     elif isinstance(e, odoo.exceptions.ValidationError):
         tmp["exception_type"] = "validation_error"
+    elif isinstance(e, odoo.exceptions.ViewError):
+        tmp["exception_type"] = "view_error"
```
Comment (rco-odoo) [inline odoo/http.py]: Ouch!  Do you really need a new category of error here?

File: odoo/models.py [hunk 12]
```diff
@@ -1367,6 +1367,10 @@ def load_views(self, views, options=None):
         :param options['action_id']: id of the action to get the filters
         :return: dictionary with fields_views, fields and optionally filters
         """
+        # populate the context with essential information for debugging problematic views
+        # all the way up the stack
+        self = self.with_context(loaded_view=[self._name, views, options])
```
Comment (rco-odoo) [inline odoo/models.py]: That sucks.  Looks like a lazy way to fix a problem :wink:

File: addons/base_automation/models/base_automation.py [hunk 13]
```diff
@@ -207,7 +220,25 @@ def _process(self, records, domain_post=None):
...
-                    self.action_server_id.with_context(**ctx).run()
+                    action = {
+                            'label': 'Fix me',
+                            'action': {
+                                'type': 'ir.actions.act_window',
+                                'res_model': 'base.automation.failure.wizard',
+                                'target': 'new',
+                                'views': [[
+                                    self.env.ref(
+                                        'base_automation.base_automation_failure_wizard_form').id,
```
[the remark this answers is not in the document]
Comment (Elkasitu) [reply addons/base_automation/models/base_automation.py]: Could be done, but it provides no context to whoever is fixing it and more importantly, it does not display the warning message saying "if you do this you'll most likely corrupt your data".

At this point I don't care about the approach as long as everyone agrees on a spec

File: addons/base_automation/models/base_automation.py [hunk 14]
```diff
@@ -309,7 +340,26 @@ def base_automation_onchange(self):
...
+                                self.env.ref(
+                                    'base_automation.base_automation_failure_wizard_form').id,
+                                'form']],
+                            },
+                        'options': {'additional_context': {'active_ids': self.ids}},
+                        'visible': self.user_has_groups('base.group_system'),
+                        }
+                try:
+                    res = server_action.run()
+                except Exception as e:
```
[the remark this answers is not in the document]
Comment (Elkasitu) [reply addons/base_automation/models/base_automation.py]: I guess by good reason you mean a business exception (UserError)? In this case IMO the action should handle it, if it does not handle an *expected* exception it's considered a badly written action

File: addons/base_automation/models/base_automation.py [hunk 15]
```diff
@@ -309,7 +340,26 @@ def base_automation_onchange(self):
...
+                            'type': 'ir.actions.act_window',
+                            'res_model': 'base.automation.failure.wizard',
+                            'target': 'new',
+                            'views': [[
+                                self.env.ref(
+                                    'base_automation.base_automation_failure_wizard_form').id,
+                                'form']],
+                            },
+                        'options': {'additional_context': {'active_ids': self.ids}},
+                        'visible': self.user_has_groups('base.group_system'),
```
[the remark this answers is not in the document]
Comment (Elkasitu) [reply addons/base_automation/models/base_automation.py]: Yes, I did this so that the burden of checking access rights is on the implementation of post-mortem actions instead of on the one who defines a post-mortem action, the `__init__` will discard it if it's not visible.

In hindsight, it might be better to just call it `groups` and pass a list of groups and then letting the `__init__` call `user_has_groups`

File: odoo/exceptions.py (hunk 7 above)
Comment (Elkasitu) [reply odoo/exceptions.py]: I had already implemented it locally, but pushed it just now

File: odoo/exceptions.py (hunk 8 above)
Comment (Elkasitu) [reply odoo/exceptions.py]: there may be places that try to render a view and check for a `ValueError`, I did it since the task is supposed to be somewhat backportable

File: odoo/exceptions.py (hunk 9 above)
Comment (Elkasitu) [reply odoo/exceptions.py]: True, but given your comment below, should we even log this and if so, is info really the right log level?

File: odoo/http.py (hunk 11 above)
Comment (Elkasitu) [reply odoo/http.py]: The web client depends on this to name the modal among other things, otherwise it's treated as a generic server error... It's not strictly necessary but I don't see why it's so bad

File: odoo/models.py (hunk 12 above)
Comment (Elkasitu) [reply odoo/models.py]: The alternative would be passing down these 3 parameters down to `fields_view_get`, `post_process_and_fields` and then to `post_process` (which will be called recursively), given that and the fact that the task is supposed to be backportable, I thought this was the best approach

File: odoo/exceptions.py [hunk 16]
```diff
@@ -107,3 +110,38 @@ def __init__(self, msg, tb):
...
+            loaded_view = view.env.context.get('loaded_view', False)
+            if loaded_view:
+                model = loaded_view[0]
+
+                try:
+                    # reload the views to retrigger a ViewError, this time with debug information
+                    # to properly find the view id that causes the error
+                    view.env[model].with_context(inherit_branding=True).load_views(*loaded_view[1:])
+                except ViewError as e:
+                    vals = [self, e.name, e.value, e.actions]
```
Comment (rco-odoo) [inline odoo/exceptions.py]: Why don't you reintroduce a private method like `view._raise_view_error(msg)` on `ir.ui.view` to:
- retry the view construction like above and simply reraise `e`, or
- log the error and raise a `ViewError` with the right action.

This would automatically factor out the logging part, and `ViewError` would have no actual code.

File: odoo/models.py [hunk 17]
```diff
@@ -1367,6 +1367,10 @@ def load_views(self, views, options=None):
         :param options['action_id']: id of the action to get the filters
         :return: dictionary with fields_views, fields and optionally filters
         """
+        # populate the context with essential information for debugging problematic views
+        # all the way up the stack
+        self = self.with_context(loaded_view=(self._name, views, options))
```
Comment (rco-odoo) [inline odoo/models.py]: Why don't you simply pass `self`?
```suggestion
        self = self.with_context(load_views_args=(self, views, options))
```

File: odoo/addons/base/models/ir_ui_view.py [hunk 18]
```diff
@@ -556,24 +571,34 @@ def get_inheriting_views_arch(self, view_id, model):
...
-            'xmlid': view.xml_id or not_avail,
-            'model': view.model or not_avail,
-            'parent': view.inherit_id.id or not_avail,
-            'msg': message,
-        }
-        _logger.info(message)
-        raise ValueError(message)
+    def raise_view_error(self, message, view, recurse=False):
+        log_view_error(message, view)
+        if not view and recurse:
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_ui_view.py]: I think that `recurse` could be derived from the context:
```suggestion
        if not view and not view.env.context.get('inherit_branding'):
```

File: odoo/addons/base/models/ir_ui_view.py [hunk 19]
```diff
@@ -556,24 +571,34 @@ def get_inheriting_views_arch(self, view_id, model):
...
+                view.env[model].with_context(inherit_branding=True).load_views(*loaded_view[1:])
+        else:
+            actions = []
+            if view.user_has_groups('base.group_system'):
+                actions = [{
+                    'label': 'Go to view',
+                    'action': {
+                        'type': 'ir.actions.act_window',
+                        'res_model': 'ir.ui.view',
+                        'res_id': view.id,
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_ui_view.py]: This code should not be reachable if `view` is empty...

**Review by rco-odoo — APPROVED** [review]
@robodoo r+

**Comment by antonylesuisse** [comment]
@robodoo  r-

File: odoo/http.py (hunk 11 above)
Comment (antonylesuisse) [reply odoo/http.py]: Avoid adding errors

File: odoo/addons/base/models/ir_ui_view.py [hunk 20]
```diff
@@ -555,25 +570,33 @@ def get_inheriting_views_arch(self, view_id, model):
...
-            'view_name': view.name or not_avail,
-            'viewid': view_id or not_avail,
-            'xmlid': view.xml_id or not_avail,
-            'model': view.model or not_avail,
-            'parent': view.inherit_id.id or not_avail,
-            'msg': message,
-        }
-        _logger.info(message)
-        raise ValueError(message)
+    def raise_view_error(self, message):
```
Comment (antonylesuisse) [inline odoo/addons/base/models/ir_ui_view.py]: Dont change the call of raise_view_error juste raise here using the view_id received.

File: odoo/exceptions.py [hunk 21]
```diff
@@ -13,24 +13,27 @@
...
+    def __init__(self, name, value=None, actions=None):
         if type(self) == except_orm:
             caller = frame_codeinfo(currentframe(), 1)
-            _logger.warn('except_orm is deprecated. Please use specific exceptions like UserError or AccessError. Caller: %s:%s', *caller)
+            _logger.warn('except_orm is deprecated. Please use specific exceptions like UserError '
+                         'or AccessError. Caller: %s:%s', *caller)
         self.name = name
         self.value = value
         self.args = (name, value)
+        self.actions = actions or []
```
Comment (antonylesuisse) [inline odoo/exceptions.py]: remove from signature change, juste add actions = False to document the feature. Dont reindent other lines.

File: odoo/models.py [hunk 22]
```diff
@@ -1385,6 +1385,10 @@ def load_views(self, views, options=None):
         :param options['action_id']: id of the action to get the filters
         :return: dictionary with fields_views, fields and optionally filters
         """
+        # populate the context with essential information for debugging problematic views
+        # all the way up the stack
+        self = self.with_context(load_view_args=(self, views, options))
```
Comment (antonylesuisse) [inline odoo/models.py]: We will use the view_id from the current code, dont wrap this in the context.

File: odoo/addons/base/models/ir_ui_view.py [hunk 23]
```diff
@@ -555,25 +570,33 @@ def get_inheriting_views_arch(self, view_id, model):
...
-        ) % {
-            'view_name': view.name or not_avail,
-            'viewid': view_id or not_avail,
-            'xmlid': view.xml_id or not_avail,
-            'model': view.model or not_avail,
-            'parent': view.inherit_id.id or not_avail,
-            'msg': message,
-        }
-        _logger.info(message)
-        raise ValueError(message)
```
Comment (antonylesuisse) [inline odoo/addons/base/models/ir_ui_view.py]: v = ValuerError(messsge)
v.actions = [{'view':view_id}]
raise v

File: odoo/exceptions.py (hunk 8 above)
Comment (antonylesuisse) [reply odoo/exceptions.py]: lets keep the changes minimal dont add a new type of exception

File: odoo/exceptions.py [hunk 24]
```diff
@@ -13,24 +13,27 @@
 import logging
 from inspect import currentframe
 from .tools.func import frame_codeinfo
+from .tools.translate import _
```
Comment (odony) [inline odoo/exceptions.py]: unused import?

File: odoo/addons/base/models/ir_ui_view.py [hunk 25]
```diff
@@ -555,25 +570,33 @@ def get_inheriting_views_arch(self, view_id, model):
...
+        log_view_error(message, self)
+        actions = []
+        if not self and not self.env.context.get('inherit_branding'):
+            # an error occurred when loading a default view, find out
+            # especifically which one broke
+            args = self.env.context.get('load_view_args')
+            if args:
+                # reload the view to trigger a ViewError, this time with
+                # branding to properly find the view that causes the error
+                args[0].with_context(inherit_branding=True).load_views(*args[1:])
```
Comment (odony) [inline odoo/addons/base/models/ir_ui_view.py]: Does this second call to `load_views` still serve a purpose? It does not seem to do what the comment says, as nothing apparently needs/uses the branding. The various calls to raise_view_error() were already providing the leaf view that triggers the error, no?
It still seems to work and point to the leaf view when I remove the second call, too.

**Comment by Elkasitu** [comment]
ping @d-fence

File: odoo/addons/base/models/ir_ui_view.py (hunk 23 above)
Comment (Elkasitu) [reply odoo/addons/base/models/ir_ui_view.py]: It is not so simple! `view_id` can be `False`

File: odoo/exceptions.py (hunk 21 above)
Comment (Elkasitu) [reply odoo/exceptions.py]: why not keep it in master and then make a backportable PR?

File: odoo/addons/base/models/ir_ui_view.py (hunk 25 above)
Comment (Elkasitu) [reply odoo/addons/base/models/ir_ui_view.py]: Yes, if you look at the places where `raise_view_error` is called, you'll see that in the method `post_process_and_fields` there are two instances where we get the `data-oe-id` attribute from an element in the view corresponding to a non-existing field.

This is because you can load a view without providing a `view_id`, if this is the case and any inherited view references a non-existing field, `load_views` will crash but it won't have any way of knowing *which* inherited view triggered the crash.

Basically, this is not needed for most cases, but it is the case that @dbo-odoo needed covered (see task chatter)

File: odoo/addons/base/models/ir_ui_view.py (hunk 20 above)
Comment (Elkasitu) [reply odoo/addons/base/models/ir_ui_view.py]: the idea was to make a PR for master and another for stable with minimal modifications, but OK

File: addons/base_automation/models/base_automation.py [hunk 26]
```diff
@@ -26,6 +26,19 @@
 }
 
 
+class AutomationFailureWizard(models.TransientModel):
+    _name = 'base.automation.failure.wizard'
+    _description = 'Automated action failure wizard'
+
+    def deactivate(self):
+        failures = self.env['base.automation'].browse(self._context.get('active_ids'))
```
[the remark this answers is not in the document]
Comment (Elkasitu) [reply addons/base_automation/models/base_automation.py]: this commit was not intended to be back-portable, I think it's dangerous to simply allow users to disable automated actions which might render their future data corrupted without displaying a big warning message...

Perhaps for stable we could simply add a button which will open the backend form view of the `base.automation`

OK for your first comment

[the remark this answers is not in the document]
Comment (odony) [reply addons/base_automation/models/base_automation.py]: > this commit was not intended to be back-portable, I think it's dangerous to simply allow users to disable automated actions which might render their future data corrupted without displaying a big warning message...

Of course, and the feature itself is very risky for this reason. What I meant is that the "big warning message"  would have to be implemented client-side in JS, not with a server-side wizard, if you want to be able to backport it. 
Normally we would recommend to put as much business logic as possible on the server-side, and keep the presentation client-side. But it does not work if you want to target stable series. And, after all, this can be seen as a kind of extension of the crash manager.

**Review by bouvyd — COMMENTED** [review]
i didn't want to have this review request pending forever.

i have no opinion regarding this, i'll leave it to the framework team to handle it

bisous
